rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+	repeated string provided_capabilities = 8;
+
+	// Whether or not the plugin is enabled. An unset value 
+	// should be considered as true.
+	boolean enabled = 9;
+}
+```
+
+### Protobuf as the Remote Contract
+As a departure from PF4J, plugin contracts will be defined as Protobuf messages (not services), so long as they are _remote capable_. If a plugin is in-process only, and there’s no expectation that it becomes available for remote plugins, it does not need to define its contract as protobuf.
+
+However, for any remotely-capable extension, its contract must be defined as protobufs, rather than exposed as simple POJOs or interfaces. A key distinction is that we’re not saying gRPC: A remote transport may be Amazon SQS or Redis PubSub, not necessarily gRPC.
+
+### Technology Compatibility Kits
+_alias: TCKs_
+
+A convention of writing TCKs for each Extension Point will be established. The concept of a TCK is borrowed from the JSR world, where compatibility tests are written to ensure that various JREs must implement to assert correct behavior of various features. For Spinnaker plugins, TCKs can be written for two different consumers: Extension Point providers and Plugin developers.
+
+TCKs satisfy asserting correct behavior of an extension point and its associated plugin(s). For any extension point, a TCK will be written that asserts even pathological plugins behave in an expected way while integrating with the extension point. For a plugin, a TCK will be provided that a plugin developer can use to assert correctness without necessarily needing to run other Spinnaker services.
+
+This goes further than the convention of using Protobuf to define the interfaces of a particular plugin, where Protobufs only define the interface, whereas a TCK will assert behavior that lies beneath the interface. Combining a TCK and adherence to the interface, contracts will be fully asserted between a plugin and its extension point(s).
+
+### Up/Downgrading Plugins
+This section applies to federated development scenarios, where the operator is not necessarily the same as the plugin developers; where the release cadence of a plugin does not adhere to the release cadence of the core Spinnaker services.
+
+The upgrading (or downgrading, from now on broadly upgrading) plugin story is different based on the particular plugin runtime.
+
+#### In-process
+When a service starts up, PF4J will scan a designated locally-accessible directory for plugins to install (these plugins may be downloaded at runtime or baked into the image at built-time). Once an application has started, it will notify Front50 of the plugins that are installed, and get a response dictating which plugins are to be enabled.
+
+In the case of a runtime-downloaded plugin, version to download will be dictated by Front50. Should Front50 be unavailable, the service will download the version defined by the service, or if none is defined, the latest plugin version that matches the Spinnaker service version.
```

Comment:
is this complexity needed? Why not, at least in the beginning, say no front50 - no plugins.

Reply:
Because there's the possibility that Front50 may be down. This isn't often, but if plugins existed that provided critical functionality, and then while a service was starting, Front50 was unavailable for a moment, we'd lose the critical functionality entirely.

Reply:
hm... if it's to account for momentary blips in front50 wouldn't it be better to retry/fail service start if we can't load the critical plugins rather than potentially loading a potentially wrong version of the plugin?

Reply:
I'm going to follow the flow directed by https://github.com/spinnaker/community/pull/43#discussion_r333649332

Would this address your feedback?

Reply:
as a spinnaker operator, if I expect a plugin to be loaded and its not, I'd like the deploy to fail or retry, as the desired state has not been achieved. This is particularly important and would be confusing if I deploy a new version of a plugin and the new functionality is not present.

Reply:
Yes, agreed.